### PR TITLE
Add separate keyword argument dictionaries for components of boxenplot

### LINKED ISSUE #2701: Boxenplot fliersize ?
Hi, is it possible to change `fliersize` in `boxenplot`, as in `boxplot`?

### PR DESCRIPTION
**Pull request by EitanHemed**
Aiming to solve issue #2701  - adding separate keyword support for `boxenplot`.

Changes - 
1. Removed `kwargs` which were used to draw the median lines and scatter plot of outliers previously.

2. Added separate kwargs - `box_kws`, `line_kws` (drawing the median lines) and `flier_kws` (for the scatter of outliers).

3. Updated the matching docstring.

Passed all relevant tests. 

Closes #2701

### CODE REVIEW

File: seaborn/categorical.py [hunk 1]
```diff
@@ -1840,24 +1840,51 @@ def _width_functions(self, width_func):
...
 
         vert = self.orient == "v"
         x = positions[0]
         box_data = np.asarray(box_data)
 
         # If we only have one data point, plot a line
         if len(box_data) == 1:
-            kws.update({
-                'color': self.gray, 'linestyle': '-', 'linewidth': self.linewidth
+            line_kws.update({
```
Comment (mwaskom) [inline seaborn/categorical.py:1877]: I think this should be sourcing kwargs from `box_kws`, after you resolve the default/custom options:

- `edgecolor` -> `color`
- `linewidth` -> `linewidth`
- `linestyle` -> `linestyle`

Comment (mwaskom) [reply seaborn/categorical.py:1877]: (I am not sure about how to handle the corner case where the box linewidth is 0, which would be valid for boxes but make this artist invisible. Perhaps `linewidth=max(box_kws["linewidth"], line_kws["linewidth"])`?

File: seaborn/categorical.py [hunk 2]
```diff
@@ -1840,24 +1840,51 @@ def _width_functions(self, width_func):
...
+        # Set the default kwargs for the outliers scatterplot
+        flier_default_kws = dict(marker='d', color=self.gray)
+
+        # Update each of the dicts containing box/line/fliers kwargs with the
+        # default values (if not defined by the user)
+        default_dicts = [line_default_kws, flier_default_kws]
+        user_input_dicts = [line_kws, flier_kws]
+        for default_dict, user_dict in zip(default_dicts, user_input_dicts):
+            for k, v in default_dict.items():
+                user_dict.setdefault(k, v)
```
Comment (mwaskom) [inline seaborn/categorical.py:1869]: Similar to the comment above, this doesn't actually reduce the amount of code vs two separate for-loops, it just makes it a little bit harder to follow.

File: seaborn/categorical.py [hunk 3]
```diff
@@ -2423,26 +2461,30 @@ def boxenplot(
...
+    line_kws: key, value mappings
+        Other keyword arguments are passed through to
+        :meth:`matplotlib.axes.Axes.plot`.
+    flier_kws: line_kws: key, value mappings
+        Other keyword arguments are passed through to
+        :meth:`matplotlib.axes.Axes.scatter`.
     kwargs : key, value mappings
         Other keyword arguments are passed through to
         :meth:`matplotlib.axes.Axes.plot` and
         :meth:`matplotlib.axes.Axes.scatter`.
```
Comment (mwaskom) [inline seaborn/categorical.py:2476]: Please remove.

**Review by mwaskom — CHANGES_REQUESTED** [review]
Thanks @EitanHemed for taking a crack at this. The logic looks right to me; I have a number of comments that are mostly cosmetic.

The comments about the `boxenplot` docstring are critical though and I am wondering how those changes ended up in here ... are you using some kind of automated formatting tool that introduced them? If so, that tools seems broken.

More importantly, could you please add a test to cover this new behavior? Thanks.

File: seaborn/categorical.py [hunk 4]
```diff
@@ -2423,26 +2461,30 @@ def boxenplot(
     showfliers : bool, optional
         If False, suppress the plotting of outliers.
     {ax_in}
+    box_kws: key, value mappings
+        Other keyword arguments are passed through to
+        :meth:`matplotlib.patches.Rectangle`.
```
Comment (mwaskom) [inline seaborn/categorical.py:2466]: This should be `:class:`, not `:meth:`.

File: seaborn/categorical.py [hunk 5]
```diff
@@ -2423,26 +2461,30 @@ def boxenplot(
     showfliers : bool, optional
         If False, suppress the plotting of outliers.
     {ax_in}
+    box_kws: key, value mappings
```
Comment (mwaskom) [inline seaborn/categorical.py:2464]: This is a `dict` now.

File: seaborn/categorical.py [hunk 6]
```diff
@@ -2423,26 +2461,30 @@ def boxenplot(
...
+        Other keyword arguments are passed through to
```
Comment (mwaskom) [inline seaborn/categorical.py:2465]: This isn't "other keyword arguments", please rephrase as `"Keyword arguments for the box artists; passed to ..."`

File: seaborn/categorical.py [hunk 7]
```diff
@@ -2423,26 +2461,30 @@ def boxenplot(
...
+        :meth:`matplotlib.patches.Rectangle`.
+    line_kws: key, value mappings
+        Other keyword arguments are passed through to
```
Comment (mwaskom) [inline seaborn/categorical.py:2468]: Same comment as above, but please clarify that this pertains to the line denoting the median.

File: seaborn/categorical.py [hunk 8]
```diff
@@ -2423,26 +2461,30 @@ def boxenplot(
...
+        :meth:`matplotlib.axes.Axes.plot`.
+    flier_kws: line_kws: key, value mappings
+        Other keyword arguments are passed through to
```
Comment (mwaskom) [inline seaborn/categorical.py:2471]: Same comment as above, but please clarify that this pertains to the outliers.

File: seaborn/categorical.py [hunk 9]
```diff
@@ -1912,33 +1939,41 @@ def horz_perc_box(x, b, i, k, w):
...
             if len(outliers) > 0:
-                ax.scatter(xs_outliers, ys_outliers, marker='d',
-                           c=self.gray, **kws)
+                ax.scatter(xs_outliers, ys_outliers,
+                           **flier_kws
+                           )
 
             # Construct a color map from the input color
             rgb = [hex_color, (1, 1, 1)]
             cmap = mpl.colors.LinearSegmentedColormap.from_list('new_map', rgb)
```
[the remark this answers is not in the document]
Comment (EitanHemed) [reply seaborn/categorical.py:1957]: I'd be happy to, but i i'll remove them a NameError will be raised on the following line of code. 
`rgb = [hex_color, cmap(.85)]` requires cmap to be defined.  

I'll take a look after resolving the rest of the issues with the pull request.

File: seaborn/categorical.py [hunk 10]
```diff
@@ -1840,24 +1840,51 @@ def _width_functions(self, width_func):
...
-                widths=1, ax=None, **kws):
+                widths=1, ax=None, box_kws=None,
+                flier_kws=None,
+                line_kws=None):
+
+        # -- Default keyword dicts - based on
+        # distributions.plot_univariate_histogram
+        box_kws, flier_kws, line_kws = map(
+            lambda d: {} if d is None else d.copy(),
+            (box_kws, flier_kws, line_kws))
```
[the remark this answers is not in the document]
Comment (EitanHemed) [reply seaborn/categorical.py:1851]: Sure thing.

File: seaborn/categorical.py [hunk 11]
```diff
@@ -1912,33 +1939,41 @@ def horz_perc_box(x, b, i, k, w):
...
             rgb = [hex_color, cmap(.85)]
             cmap = mpl.colors.LinearSegmentedColormap.from_list('new_map', rgb)
+
+            # Set the default kwargs for the boxes
+            box_default_kws = dict(cmap=cmap,
+                                   edgecolor=self.gray,
+                                   linewidth=self.linewidth)
+
+            for k, v in box_default_kws.items():
+                box_kws.setdefault(k, v)
```
[the remark this answers is not in the document]
Comment (EitanHemed) [reply seaborn/categorical.py:1968]: I moved most of it upwards, but kept the `cmap` argument below, as it is only being assigned on this part of the function. 

`cmap` itself cannot be easily moved upwards, as it relies on `rgb`, which relies on `hex_color`.

File: seaborn/categorical.py (hunk 2 above)
Comment (EitanHemed) [reply seaborn/categorical.py:1869]: OK, thank you for your feedback.

File: seaborn/categorical.py (hunk 1 above)
Comment (EitanHemed) [reply seaborn/categorical.py:1877]: The solution seems good to me. 

`linestyle` is not defined by default on box_kws, hence the added `box.kws` call.

File: seaborn/categorical.py [hunk 12]
```diff
@@ -2416,8 +2406,9 @@ def boxenplot(
     return ax
 
 
-boxenplot.__doc__ = dedent("""\
+boxenplot.__doc__ = dedent("""
```
Comment (mwaskom) [inline seaborn/categorical.py:2409]:
```suggestion
boxenplot.__doc__ = dedent("""\
```

**Comment by mwaskom** [comment]
Thanks @EitanHemed I took care of the final touches (fixing the lint errors and test failures on pinned versions) so that this could make it into the release candidate.

**Comment by EitanHemed** [comment]
@mwaskom great, thanks!

**Pull request closed without merging**